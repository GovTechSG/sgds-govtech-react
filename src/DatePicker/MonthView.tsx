import classNames from 'classnames';
import * as React from 'react';

export interface MonthViewProps extends React.HTMLAttributes<HTMLElement> {
  onClickMonth : Function;
  displayDate: Date;
}

export const MONTH_LABELS = [
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec',
];

export const MonthView = React.forwardRef<HTMLDivElement, MonthViewProps>(
  ({ displayDate, onClickMonth, ...props }, ref) => {
    const displayMonth = MONTH_LABELS[displayDate.getMonth()];

    return (
      <div className="container text-center" ref={ref} {...props}>
        <div className="row">
          {MONTH_LABELS.map((m, idx) => (
            <div
              className={classNames(displayMonth === m  && 'active', 'col-4')}
              key={m}
            >
              <button onClick={() =>onClickMonth(idx)}>{m}</button>
            </div>
          ))}
        </div>
      </div>
    );
  }   
);

export default MonthView;
