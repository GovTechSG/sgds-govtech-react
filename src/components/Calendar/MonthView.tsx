import classNames from 'classnames';
import * as React from 'react';
import Button from '../Button/Button';
import Col from '../Col';
import DatePickerContext from './DatePickerContext'
import { useState, useContext } from 'react';
import { CalendarState } from './DatePicker';
export interface MonthViewProps extends React.HTMLAttributes<HTMLElement> {
  displayDate: Date;
  state: CalendarState;
  setState: React.Dispatch<React.SetStateAction<CalendarState>>
}


const MONTH_LABELS = [
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec',
];

export const MonthView = React.forwardRef<HTMLDivElement, MonthViewProps>(
  ({ displayDate, state, setState, ...props }, ref) => {
    const displayMonth = MONTH_LABELS[displayDate.getMonth()];
    const {setView}  = useContext(DatePickerContext)

    const handleClickMonth = (month: number) => {
      setView('day')
      setState({
        ...state,
        displayDate: new Date(state.displayDate.setMonth(month))
      })
    }
    return (
      <div className="container text-center" ref={ref} {...props}>
        <div className="row">
          {MONTH_LABELS.map((m, idx) => (
            <div
              className={classNames(displayMonth === m && 'active', 'col-4')}
              key={m}
            >
              <button onClick={() =>handleClickMonth(idx)}>{m}</button>
            </div>
          ))}
        </div>
      </div>
    );
  }
);

export default MonthView;
