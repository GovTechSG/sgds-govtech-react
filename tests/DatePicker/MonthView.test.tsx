import * as React from 'react';
import { fireEvent, render } from '@testing-library/react';
import MonthView, {MONTH_LABELS} from '../../src/DatePicker/MonthView'

describe('MonthView', () => {
    const mockFn = jest.fn()
    const displayDate = new Date()
    it ('should have default html structure', ()=> {
      
        const {container} = render(<MonthView onClickMonth={mockFn} displayDate={displayDate} />)

        expect(container.querySelector('.sgds.monthpicker')).toBeInTheDocument()
        expect(container.querySelectorAll('button.month').length).toEqual(12)
        expect(container.querySelectorAll('button.active.month').length).toEqual(1) 
        expect(container.querySelector('button.active')?.textContent).toEqual(MONTH_LABELS[displayDate.getMonth()]) 

    }) 

    it('onClick handler should work', () => {
        const { container} = render(<MonthView onClickMonth={mockFn} displayDate={displayDate} />)
        expect(container.querySelectorAll('button').length).toEqual(12)

        const buttonOne = container.querySelectorAll('button')[0]

        fireEvent.click(buttonOne)

        expect(mockFn).toHaveBeenCalled() 
    })
})