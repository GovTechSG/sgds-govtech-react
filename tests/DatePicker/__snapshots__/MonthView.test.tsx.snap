// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthView should have default html structure 1`] = `
<DocumentFragment>
  <div
    class="container text-center"
  >
    <div
      class="row"
    >
      <div
        class="col-4"
      >
        <button>
          Jan
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          Feb
        </button>
      </div>
      <div
        class="active col-4"
      >
        <button>
          Mar
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          Apr
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          May
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          Jun
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          Jul
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          Aug
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          Sep
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          Oct
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          Nov
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          Dec
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
